### PR TITLE
[Networking] GossipSub Replay Message Mitigation

### PR DESCRIPTION
**Pull request by kc1116**
This PR introduces measures to mitigate GossipSub message replay attacks. It overrides the [DuplicateMessage](https://github.com/libp2p/go-libp2p-pubsub/blob/master/trace.go#L48) callback in the pubsub tracer of the node, allowing the storage of a counter for each peer to track duplicate messages detected.

A new callback, `DuplicateMessageCount`, has been added to the mesh tracer. This callback is utilized in the score registry to retrieve the current count of duplicate messages detected for a peer. The count is then multiplied by `DefaultDuplicateMessagePenalty` and applied to the application-specific score.

To facilitate this, a cache is implemented to store the count for each peer. The count in the cache will decay over time, ensuring that peers are not penalized indefinitely, and the count eventually decays to 0 during sustained good behavior.

ref: https://github.com/dapperlabs/flow-go/issues/6809

### CODE REVIEW

File: network/p2p/scoring/registry.go [hunk 1]
```diff
@@ -401,12 +417,24 @@ func (r *GossipSubAppSpecificScoreRegistry) subscriptionPenalty(pid peer.ID, flo
 	return 0
 }
 
+// duplicateMessagesPenalty returns the duplicate message penalty for a peer.
+func (r *GossipSubAppSpecificScoreRegistry) duplicateMessagesPenalty(pid peer.ID) float64 {
+	duplicateMessageCount := r.getDuplicateMessageCount(pid)
+	if duplicateMessageCount != 0 {
```
Comment (yhassanzadeh13) [inline network/p2p/scoring/registry.go:423]: In addition to `DefaultDuplicateMessagePenalty = -10e-4`, we should also define a `DefaultDuplicateMessageThreshold = 10e+4`. Based on the current approach on tracking the duplicate message delivery using the GossipSub tracer, the counter value does not necessarily indicate a misbehavior itself. The GossipSub will count every message that is _not fresh_ as a duplicate. So all gossips for the same message except the first one are considered duplicates. 

We should treat duplicate message count of a node as a misbehavior only when it stands as an anomaly. To decide whether the count is anomaly or not, we define the threshold, and we set to a high value (10+e4). Accordingly, here, we should check whether the `duplicateMessageCount > threshold`, and if it is, we should penalize the node for the _excess_ amount floored at the `MaxAppSpecificPenalty` (as it is now). 

With a regular decay of 0.5, the small penalty of `-10e-4` and a large threshold of `10e4`, the normal behavior of the nodes are set safe, while anomalies are treated as misbehavior and penalized with the excess.

Please also add test ensuring that as long as the node stays below the threshold on duplicate message delivery, it is not penalized.

File: network/p2p/tracer/internal/duplicate_msgs_counter_cache.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+	record := duplicateMessagesCounter(entity)
+	record.Gauge++
+	record.lastUpdated = time.Now()
+	// Return the adjusted record.
+	return record
+}
+
+// decayAdjustment performs geometric recordDecay on the duplicate message counter gauge of a peer. This ensures a peer is not penalized forever.
+func (g *GossipSubDuplicateMessageTrackerCache) decayAdjustment(counter DuplicateMessagesCounterEntity) (DuplicateMessagesCounterEntity, error) {
+	received := counter.Gauge
```
Comment (yhassanzadeh13) [inline network/p2p/tracer/internal/duplicate_msgs_counter_cache.go:150]:
```suggestion
	received := counter.Value
```
counter.Gauge may seem a bit confusing.

File: network/p2p/tracer/internal/duplicate_msgs_counter_cache.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+	if received < skipDecayThreshold {
+		counter.Gauge = 0
+		return counter, nil
+	}
+
+	decayedVal, err := scoring.GeometricDecay(received, g.recordDecay, counter.lastUpdated)
+	if err != nil {
+		return counter, fmt.Errorf("could not recordDecay duplicate message counter gauge: %w", err)
+	}
+	counter.Gauge = decayedVal
```
Comment (yhassanzadeh13) [inline network/p2p/tracer/internal/duplicate_msgs_counter_cache.go:164]: Here, we may be safer to make a sanity check:
```go
	if decayedVal > received {
		return counter, fmt.Errorf("unexpected recordDecay value %f for duplicate message counter gauge %f", decayedVal, received)
	}
```

File: network/p2p/scoring/registry_test.go [hunk 4]
```diff
@@ -385,6 +385,112 @@ func testScoreRegistrySpamRecordWithSubscriptionPenalty(t *testing.T, messageTyp
...
+	reg.OnInvalidControlMessageNotification(&p2p.InvCtrlMsgNotif{
+		PeerID:  peerID,
+		MsgType: messageType,
+	})
+
+	// the penalty should now be updated in the spamRecords
+	record, err, ok := spamRecords.Get(peerID) // get the record from the spamRecords.
+	assert.True(t, ok)
+	assert.NoError(t, err)
+	assert.Less(t, math.Abs(expectedPenalty-record.Penalty), 10e-3)        // penalty should be updated to -10.
```
Comment (yhassanzadeh13) [inline network/p2p/scoring/registry_test.go:472]: Please use `unittest.RequireNumericallyClose` instead, for this and other repetitions of this.

File: network/p2p/tracer/gossipSubMeshTracer.go [hunk 5]
```diff
@@ -196,6 +205,23 @@ func (t *GossipSubMeshTracer) SendRPC(rpc *pubsub.RPC, _ peer.ID) {
...
 }
 
+// DuplicateMessage is invoked when a duplicate message is dropped.
+func (t *GossipSubMeshTracer) DuplicateMessage(msg *pubsub.Message) {
+	peerID := msg.ReceivedFrom
+	count, err := t.duplicateMessageTrackerCache.Inc(msg.ReceivedFrom)
+	if err != nil {
+		t.logger.Err(err).
+			Bool(logging.KeyNetworkingSecurity, true).
+			Str("peer_id", peerID.String()).
```
Comment (yhassanzadeh13) [inline network/p2p/tracer/gossipSubMeshTracer.go:215]: Please use `p2plogging.PeerId(peerID)` for this as well as other repetitions that a peer id is logged. This logger utilizes an underlying cache that addresses a known issue with encoding peer ids.

File: network/p2p/tracer/gossipSubMeshTracer.go [hunk 6]
```diff
@@ -196,6 +205,23 @@ func (t *GossipSubMeshTracer) SendRPC(rpc *pubsub.RPC, _ peer.ID) {
 	}
 }
 
+// DuplicateMessage is invoked when a duplicate message is dropped.
+func (t *GossipSubMeshTracer) DuplicateMessage(msg *pubsub.Message) {
+	peerID := msg.ReceivedFrom
+	count, err := t.duplicateMessageTrackerCache.Inc(msg.ReceivedFrom)
+	if err != nil {
+		t.logger.Err(err).
```
Comment (yhassanzadeh13) [inline network/p2p/tracer/gossipSubMeshTracer.go:213]: shouldn't this error be irrecoverable?

File: network/p2p/tracer/internal/duplicate_msgs_counter_cache_test.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+	require.NoError(t, err)
+	require.Equal(t, 1.0, count)
+	count, err = cache.Get(peerID)
+	require.NoError(t, err)
+	unittest.RequireNumericallyClose(t, 1.0, count, .1)
+}
+
+// TestGossipSubDuplicateMessageTrackerCache tests the concurrent functionality of the GossipSubDuplicateMessageTrackerCache.
+// It ensures that the Inc and Get methods work as expected, does not test counter decay specifically.
+func TestGossipSubDuplicateMessageTrackerCache_Concurrent_Inc_Get(t *testing.T) {
```
Comment (yhassanzadeh13) [inline network/p2p/tracer/internal/duplicate_msgs_counter_cache_test.go:39]: Please also add a test for concurrently incrementing the same peer id.

File: network/p2p/tracer/internal/duplicate_msgs_counter_cache_test.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+	// expect count to initially be 0
+	count, err := cache.Get(peerID)
+	require.NoError(t, err)
+	require.Equal(t, 0.0, count)
+
+	// increment the counter for a peer
+	count, err = cache.Inc(peerID)
+	require.NoError(t, err)
+	require.Equal(t, 1.0, count)
+
```
Comment (yhassanzadeh13) [inline network/p2p/tracer/internal/duplicate_msgs_counter_cache_test.go:117]:
```suggestion
	require.Eventually(t, func() bool {
		count, err = cache.Get(peerID)
		require.NoError(t, err)
		return count > 0 && count < 0.6
	}, 10*time.Second, 100*time.Millisecond)
```

File: network/p2p/scoring/registry.go [hunk 9]
```diff
@@ -401,12 +417,24 @@ func (r *GossipSubAppSpecificScoreRegistry) subscriptionPenalty(pid peer.ID, flo
...
+		duplicateMessagePenalty := duplicateMessageCount * DefaultDuplicateMessagePenalty
+		if duplicateMessagePenalty < MaxAppSpecificPenalty {
+			return MaxAppSpecificPenalty
+		}
+		return duplicateMessagePenalty
```
Comment (yhassanzadeh13) [inline network/p2p/scoring/registry.go:428]: Please collect this in a histogram.

File: network/p2p/tracer/internal/duplicate_msgs_counter_cache_test.go (hunk 7 above)
Comment (kc1116) [reply network/p2p/tracer/internal/duplicate_msgs_counter_cache_test.go:41]: https://github.com/onflow/flow-go/pull/5058/commits/07269a38d2d7edc5b906c52f09532269c851560e

File: network/p2p/tracer/gossipSubMeshTracer.go (hunk 6 above)
Comment (kc1116) [reply network/p2p/tracer/gossipSubMeshTracer.go]: https://github.com/onflow/flow-go/pull/5058/commits/1b8cf130d20cb44b9f55f3b28afb50f1a1a45e14

File: network/p2p/scoring/registry_test.go (hunk 4 above)
Comment (kc1116) [reply network/p2p/scoring/registry_test.go]: https://github.com/onflow/flow-go/pull/5058/commits/ee5d747ed7b2d21c69055b10e10bfa151124a5ea

File: network/p2p/tracer/internal/duplicate_msgs_counter_cache.go (hunk 3 above)
Comment (kc1116) [reply network/p2p/tracer/internal/duplicate_msgs_counter_cache.go]: https://github.com/onflow/flow-go/pull/5058/commits/ec997a6c41b4a4a72ed53cb936121896407188af

File: network/p2p/tracer/internal/duplicate_msgs_counter_cache.go (hunk 2 above)
Comment (kc1116) [reply network/p2p/tracer/internal/duplicate_msgs_counter_cache.go]: https://github.com/onflow/flow-go/pull/5058/commits/57fc14ab8d0e531bc517a2d09e6373fd51b13584

File: network/p2p/scoring/registry.go (hunk 1 above)
Comment (kc1116) [reply network/p2p/scoring/registry.go]: https://github.com/onflow/flow-go/pull/5058/commits/59462cd9a69895cfdb26af133d3d8d7aafe64db3

File: network/p2p/scoring/registry.go [hunk 10]
```diff
@@ -327,6 +334,15 @@ func (r *GossipSubAppSpecificScoreRegistry) computeAppSpecificScore(pid peer.ID)
...
 
+	// (5) duplicate messages penalty: the duplicate messages penalty is applied to the application specific penalty as long
+	// as the number of duplicate messages detected for a peer is greater than 0. This counter is decayed overtime, thus sustained
+	// good behavior should eventually lead to the duplicate messages penalty applied being 0.
+	duplicateMessagesPenalty := r.duplicateMessagesPenalty(pid)
+	if duplicateMessagesPenalty < 0 {
+		lg = lg.With().Float64("duplicate_messages_penalty", duplicateMessagesPenalty).Logger()
+		appSpecificScore += duplicateMessagesPenalty
+	}
+
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/scoring/registry.go]: https://github.com/onflow/flow-go/pull/5058/commits/d092114ed1ed7fef2cdabe8d9d46c93e6866147c

File: network/p2p/tracer/internal/duplicate_msgs_counter_cache.go [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+// returned.
+// Args:
+// - peerID: the peer ID of the peer in the GossipSub protocol.
+//
+// Returns:
+// - float64: updated value for the duplicate message tracker.
+// - bool: true if the counter was initialized, false otherwise.
+// - error: if any error was encountered during record adjustments in cache.
+// No errors are expected during normal operation.
+func (g *GossipSubDuplicateMessageTrackerCache) Get(peerId peer.ID) (float64, error) {
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/tracer/internal/duplicate_msgs_counter_cache.go]: https://github.com/onflow/flow-go/pull/5058/commits/d9467b0bfbc244aa5bd13efea00a202226a5ffe2

File: network/p2p/scoring/registry.go (hunk 9 above)
Comment (kc1116) [reply network/p2p/scoring/registry.go:429]: Will using the peerID as a label cause a strain in our prometheus instances since we can have any number of peerIDs? We currently have a log message that has all penalty information for a peer when the app specific score is computed, will the histogram give us some different insight?

**Comment by kc1116** [comment]
Updated cache to use new AdjustWithInit / GetWithInit pattern https://github.com/onflow/flow-go/pull/5058/commits/a62d9eeb2735e1818e196bd74c5a3347f7728b6d

File: network/p2p/tracer/internal/duplicate_msgs_counter_cache.go [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+	entityId := makeId(peerID)
+	adjustedEntity, adjusted := d.c.AdjustWithInit(entityId, adjustFunc, func() flow.Entity {
+		return NewDuplicateMessagesCounter(entityId)
+	})
+
+	if err != nil {
+		return 0, fmt.Errorf("unexpected error while applying decay and increment adjustments for peer %s: %w", peerID, err)
+	}
+
+	if !adjusted {
```
Comment (peterargue) [inline network/p2p/tracer/internal/duplicate_msgs_counter_cache.go:78]: what would cause this to be false?

File: network/p2p/tracer/gossipSubMeshTracer.go [hunk 13]
```diff
@@ -450,6 +479,22 @@ func (t *GossipSubMeshTracer) LastHighestIHaveRPCSize() int64 {
...
 
+// DuplicateMessageCount returns the current duplicate message count for the peer.
+func (t *GossipSubMeshTracer) DuplicateMessageCount(peerID peer.ID) float64 {
+	count, found, err := t.duplicateMessageTrackerCache.GetWithInit(peerID)
+	if err != nil {
+		t.logger.Err(err).
+			Bool(logging.KeyNetworkingSecurity, true).
+			Str("peer_id", peerID.String()).
+			Msg("failed to get duplicate message count for peer")
+		return 0
```
Comment (peterargue) [inline network/p2p/tracer/gossipSubMeshTracer.go:490]: is it safe to return 0 on error?

File: config/default-config.yml [hunk 14]
```diff
@@ -346,6 +353,8 @@ network-config:
             # opportunistically graft peers that are not access nodes (i.e., with -1),
             # or penalized peers (i.e., with -100).
             opportunistic-graft: 101
+            # The threshold at which the duplicate message count for a peer will result in the peer being penalized
+            duplicate-message: 10e+4
```
Comment (yhassanzadeh13) [inline config/default-config.yml:357]: Please note that this parameter is not part of the `internal` category. The `internal` category includes parameters injected directly into the libp2p pubsub system during pubsub router construction. These parameters govern the GossipSub peer scoring's internal logic, which is external to the Flow code. The peer scoring logic defined in Flow is under the `application-specific` category, where you've placed `duplicate-message-penalty`. My suggestion is to rename `duplicate-message` to `duplicate-message-threshold` and position it just before `duplicate-message-penalty` as part of the node's `application-specific` score.

File: network/p2p/tracer/gossipSubMeshTracer.go [hunk 15]
```diff
@@ -450,6 +479,22 @@ func (t *GossipSubMeshTracer) LastHighestIHaveRPCSize() int64 {
 	return t.rpcSentTracker.LastHighestIHaveRPCSize()
 }
 
+// DuplicateMessageCount returns the current duplicate message count for the peer.
+func (t *GossipSubMeshTracer) DuplicateMessageCount(peerID peer.ID) float64 {
+	count, found, err := t.duplicateMessageTrackerCache.GetWithInit(peerID)
+	if err != nil {
+		t.logger.Err(err).
+			Bool(logging.KeyNetworkingSecurity, true).
+			Str("peer_id", peerID.String()).
```
Comment (yhassanzadeh13) [inline network/p2p/tracer/gossipSubMeshTracer.go:488]:
```suggestion
			Str("peer_id", p2plogging.PeerId(peerID)).
```
https://github.com/onflow/flow-go/pull/5058#discussion_r1435337437

File: network/p2p/tracer/internal/duplicate_msgs_counter_cache_test.go [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+	// ensure that only one goroutine successfully initialized the record
+	require.Equal(t, uint(1), cache.c.Size())
+}
+
+// TestDuplicateMessageTrackerCache_DuplicateMessageReceived tests the DuplicateMessageReceived method of the RecordCache.
+// The test covers the following scenarios:
+// 1. Updating a record gauge for an existing peer ID.
+// 2. Attempting to update a record gauge  for a non-existing peer ID should not result in error. DuplicateMessageReceived should always attempt to initialize the gauge.
+// 3. Multiple updates on the same record only initialize the record once.
+func TestDuplicateMessageTrackerCache_DuplicateMessageReceived(t *testing.T) {
```
Comment (yhassanzadeh13) [inline network/p2p/tracer/internal/duplicate_msgs_counter_cache_test.go:112]: A concurrent version of this test will be highly valuable.

File: network/p2p/tracer/internal/duplicate_msgs_counter_cache_test.go [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+	// when updated the value should be incremented from 1 -> 2 and slightly decayed resulting
+	// in a gauge value less than 2 but greater than 1.9
+	gauge3, err = cache.DuplicateMessageReceived(peerID3)
+	require.NoError(t, err)
+	require.LessOrEqual(t, gauge3, 2.0)
+	require.Greater(t, gauge3, 1.9)
+}
+
+// TestDuplicateMessageTrackerCache_UpdateDecay ensures that a counter value in the record cache is eventually decayed back to 0 after some time.
+func TestDuplicateMessageTrackerCache_Decay(t *testing.T) {
```
Comment (yhassanzadeh13) [inline network/p2p/tracer/internal/duplicate_msgs_counter_cache_test.go:149]: A concurrent version of this test is also imperative, i.e., calling `DuplicateMessageReceived` concurrently `x` times on the same `peerId` should result in the `GetWithInit` value of `> 1 and < x`.

File: network/p2p/scoring/registry.go (hunk 9 above)
Comment (yhassanzadeh13) [reply network/p2p/scoring/registry.go:361]: Correct, but we require a straightforward `Histogram` rather than a `HistogramVec`, hence, no `label` would be needed. We only need to track all the returned `duplicateMessagePenalties` as separate data points, which a `Histogram` empowers. This enables us to perform queries such as determining the 0.95 percentile for duplicate message penalties across all nodes. Such queries are crucial for debugging incidents that involve a drastic decrease in a node's overall score, helping us identify the contributing factors. Therefore, I would suggest please to use a simple `Histogram` and let it observe any returned values in here.

I would also recommend having a separate `Histogram` tracking the `duplicateMessageCount` as well. These two histograms are one for the count and one for the penalty will provide a close lense in how the node behaves w.r.t to the duplicate messages it receives, as well as the health of decaying mechanism. 

>Will using the peerID as a label cause a strain in our prometheus instances since we can have any number of peerIDs? 

Since we are using a `Histogram` than a `HistogramVec`, there is no cardinality beyond the number of buckets that we set for the histogram. In other words, as we don't record the `peerId`s when observing the values, there won't be any scalability issue with the metrics cardinality relative to the number of peers. 

This is a similar approach we took for [observing the application-specific score](https://github.com/onflow/flow-go/blob/master/module/metrics/gossipsub_score.go#L134) of nodes in a [`Histogram`](https://github.com/onflow/flow-go/blob/master/module/metrics/gossipsub_score.go#L43-L51).

File: network/p2p/scoring/registry_test.go [hunk 18]
```diff
@@ -404,17 +406,233 @@ func testScoreRegistrySpamRecordWithSubscriptionPenalty(t *testing.T, messageTyp
...
+	assert.True(t, ok)
+	assert.NoError(t, err)
+	unittest.RequireNumericallyClose(t, expectedPenalty, record.Penalty, 0.01)
+	assert.Equal(t, scoring.InitAppScoreRecordStateFunc(maximumSpamPenaltyDecayFactor)().Decay, record.Decay) // decay should be initialized to the initial state.
+
+	queryTime := time.Now()
+	// eventually, the app specific score should be updated in the cache.
+	require.Eventually(t, func() bool {
+		score := reg.AppSpecificScoreFunc()(peerID)
+		return math.Abs(expectedPenalty-score)/math.Max(expectedPenalty, score) < 0.001
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/p2p/scoring/registry_test.go:611]: Please also ensure that all similar cases are refactored to use the `unittest.AreNumericallyClose` method.

**Review by yhassanzadeh13 — APPROVED** [review]
🚀 Thanks for applying the comments, added a few minor comments. Please feel free to merge once the comments are applied.

File: network/p2p/scoring/registry.go (hunk 9 above)
Comment (kc1116) [reply network/p2p/scoring/registry.go]: https://github.com/onflow/flow-go/pull/5058/commits/8729c4b32088bdd73fb0044982600c2c18605f25

File: network/p2p/tracer/gossipSubMeshTracer.go (hunk 13 above)
Comment (kc1116) [reply network/p2p/tracer/gossipSubMeshTracer.go:491]: https://github.com/onflow/flow-go/pull/5058/commits/4cea2dfabc3719d05c548fc81e0018e4f442b84b

File: network/p2p/tracer/internal/duplicate_msgs_counter_cache.go [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+
+// GetWithInit returns the current number of duplicate messages received from a peer.
+// The record is initialized before the count is returned.
+// Before the counter value is returned it is decayed using the configured decay function.
+// Returns the record and true if the record exists, nil and false otherwise.
+// Args:
+// - peerID: peerID of the remote peer.
+// Returns:
+// - The duplicate messages counter value after the decay and true if the record exists, 0 and false otherwise.
+// No errors are expected during normal operation.
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/tracer/internal/duplicate_msgs_counter_cache.go]: https://github.com/onflow/flow-go/pull/5058/commits/525398670c19ac787a68aa117291c1644da9bc15

File: network/p2p/tracer/gossipSubMeshTracer.go [hunk 20]
```diff
@@ -47,28 +48,38 @@ const (
...
+	rpcSentTracker               *internal.RPCSentTracker
+	duplicateMessageTrackerCache *internal.DuplicateMessageTrackerCache
 }
 
 var _ p2p.PubSubTracer = (*GossipSubMeshTracer)(nil)
 
+type RpcSentTrackerConfig struct {
+	CacheSize            uint32 `validate:"required"`
+	WorkerQueueCacheSize uint32 `validate:"required"`
+	WorkerQueueNumber    int    `validate:"required"`
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/tracer/gossipSubMeshTracer.go]: https://github.com/onflow/flow-go/pull/5058/commits/1968cdf17e8626c49472d90f2196651a7472abd1

File: network/p2p/tracer/gossipSubMeshTracer.go [hunk 21]
```diff
@@ -47,28 +48,38 @@ const (
...
+}
+
+type DuplicateMessageTrackerCacheConfig struct {
+	CacheSize uint32  `validate:"required"`
+	Decay     float64 `validate:"required"`
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/tracer/gossipSubMeshTracer.go]: https://github.com/onflow/flow-go/pull/5058/commits/6a7567b897e71a1f0dea95b7782c39fb93e621be

File: config/default-config.yml (hunk 14 above)
Comment (kc1116) [reply config/default-config.yml]: https://github.com/onflow/flow-go/pull/5058/commits/5d20d5e7fac783fdae8ad13c819cd9f0f5b98988

File: network/p2p/tracer/internal/duplicate_msgs_counter_cache_test.go (hunk 16 above)
Comment (kc1116) [reply network/p2p/tracer/internal/duplicate_msgs_counter_cache_test.go:112]: https://github.com/onflow/flow-go/pull/5058/commits/91a1a63836ca1b5a3e70065ec022fcc07397d03a

File: network/p2p/tracer/internal/duplicate_msgs_counter_cache_test.go (hunk 17 above)
Comment (kc1116) [reply network/p2p/tracer/internal/duplicate_msgs_counter_cache_test.go:198]: Case is covered in this test https://github.com/onflow/flow-go/pull/5058/commits/91a1a63836ca1b5a3e70065ec022fcc07397d03a

**Review by peterargue — APPROVED** [review]
